components/net/fetch/methods.rs

```diff
@@ -866,6 +869,405 @@ async fn scheme_fetch(
     }
 }
 
+fn consider_local_directory_listing(
+    request: &mut Request,
+    url: ServoUrl,
+    file_path: std::path::PathBuf,
+) -> Response {
+    /*
+    If you want to be able to browse local directories, configure
+    Servo prefs so that "browser.local_directory_listing.enabled"
+    is set to true.
+    */
+    if pref!(browser.local_directory_listing.enabled) {
+        display_local_directory_listing(request, url, file_path)
+    } else {
+        Response::network_error(NetworkError::Internal(
+            "Local directory listing feature has not been enabled in preferences".into(),
+        ))
+    }
+}
+
+fn display_local_directory_listing(
+    request: &mut Request,
+    url: ServoUrl,
+    file_path: std::path::PathBuf,
+) -> Response {
+    if !request.origin.is_opaque() {
+        /*
+        Checking for an opaque origin may not fully guarantee that the request is safe
+        (knowingly requested by the local-machine owner) rather than unsafe (came from
+        third-party JavaScript for example) so more work needs to be done to consider how to
+        guarantee safety here.
+        This check should reject requests from pages fetched from the public internet over
+        HTTP[S], FTP, WS[S], but might not stop other public scheme types from generating a
+        local directory listing. (This is why this local directory listing feature is disabled
+        by default, only enabled when the "browser.local_directory_listing.enabled" preference
+        has been set to true.)
+        TODO: Work out whether more work is needed to protect access to local directory listing.
+        */
+        return Response::network_error(NetworkError::Internal(
+            "Cannot request local directory listing from non-local origin.".to_string(),
+        ));
+    }
+    let url = if !url.path().ends_with('/') {
+        // Re-read the path as a directory, so that Servo adds the
+        // forward-slash to the end of the URL (at least internally,
+        // though the URL bar does not reflect this) and loads the
+        // linked directories and files correctly when clicked.
+        if let Ok(dir_url) = Url::from_directory_path(url.path()) {
+            ServoUrl::from_url(dir_url)
+        } else {
+            return Response::network_error(NetworkError::Internal(format!(
+                "Unable to parse local directory path {}",
+                url
+            )));
+        }
+    } else {
+        url
+    };
+    let mut response = Response::new(url, ResourceFetchTiming::new(request.timing_type()));
```

Comment:
To simplify the code and the testing, we should split this function in two parts:

1. One `directory_listing_request_allowed`, that does the origin and url checks and returns true/false. If false, return the network error.
2. If true, then do the below:

```
    let mut response = Response::new(url, ResourceFetchTiming::new(request.timing_type()));
    response.headers.typed_insert(ContentType::html());
    let directory_summary = build_directory_summary(file_path);
    let page_text = build_html_directory_listing(directory_summary);
    let bytes: Vec<u8> = page_text.into_bytes();
    *response.body.lock().unwrap() = ResponseBody::Done(bytes);
    return response;
```

That way we can test the various sub functions in isolation, and ignore the file system in the tests for now(later perhaps mocking it).